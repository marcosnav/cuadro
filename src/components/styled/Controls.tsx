import styled, { StyledProps } from 'styled-components';

interface IProps {
  newOnly?: boolean;
}

const Controls = styled.section`
  align-items: center;
  background: ${(props: StyledProps<IProps>) => props.theme.WHITE};
  border-radius: 3px;
  box-shadow: 0 4px 8px 3px rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  display: flex;
  height: 68px;
  justify-content: space-between;
  margin-top: 26px;
  padding: ${(props: StyledProps<IProps>) => props.newOnly ? '0 10px' : '0 16px'};
  transition: all 0.6s ease;
  width: ${(props: StyledProps<IProps>) => props.newOnly ? 'auto' : '300px'};
`;

Controls.displayName = 'Controls';

export default Controls;
