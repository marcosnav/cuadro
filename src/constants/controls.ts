/**
 * Available control types
 */
export enum ControlType {
  NEW_GAME = 'NEW_GAME',
  RESTART_GAME = 'RESTART_GAME',
  SEE_ORIGINAL = 'SEE_ORIGINAL',
}
