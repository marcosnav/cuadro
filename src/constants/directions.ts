/**
 * SwipeDirection, one of UP, RIGHT, LEFT, DOWN
 */
export enum SwipeDirection {
  UP = 'UP',
  RIGHT = 'RIGHT',
  DOWN = 'DOWN',
  LEFT = 'LEFT',
}
