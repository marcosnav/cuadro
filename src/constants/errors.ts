export enum errors {
  NO_INTERNET = 'NO_INTERNET',
  UNEXPECTED_ERROR = 'UNEXPECTED_ERROR',
  UNRESPONSIVE_IMAGE_PROVIDER = 'UNRESPONSIVE_IMAGE_PROVIDER',
}
