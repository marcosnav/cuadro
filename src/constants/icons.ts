import check from './../assets/svg/check.svg';
import image from './../assets/svg/image.svg';
import newicon from './../assets/svg/new.svg';
import refresh from './../assets/svg/refresh.svg';

export const icons: { [key: string]: string } = {
  check,
  image,
  new: newicon,
  refresh,
};
