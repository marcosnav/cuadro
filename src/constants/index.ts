export * from './status';
export * from './errors';
export * from './icons';
export * from './theme';
export * from './controls';
export * from './directions';
export * from './keycodes';
