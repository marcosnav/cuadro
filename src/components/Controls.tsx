import React, { FC } from 'react';
import styled from 'styled-components';
import ControlButton, { ControlType } from './ControlButton';

const ControlsWrapper = styled.section`
  align-items: center;
  background: ${(props) => props.theme.WHITE};
  border-radius: 3px;
  box-shadow: 0 4px 8px 3px rgba(0, 0, 0, 0.2);
  box-sizing: border-box;
  display: flex;
  height: 68px;
  justify-content: space-between;
  margin-top: 26px;
  padding: ${(props: { newOnly?: boolean }) => props.newOnly ? '0 10px' : '0 16px'};
  transition: all 0.6s ease;
  width: ${(props: { newOnly?: boolean }) => props.newOnly ? 'auto' : '300px'};
`;

interface IProps {
  newOnly?: boolean;
}

const secondaryControls = (onRestart: () => void, onSeeOriginal: () => void) => {
  return (
    [(
      <ControlButton
        key='restart'
        action={onRestart}
        type={ControlType.RESTART_GAME}
      >
        Restart
      </ControlButton>
    ),
    (
      <ControlButton
        key='seeoriginal'
        action={onSeeOriginal}
        type={ControlType.SEE_ORIGINAL}
      >
        See Original
      </ControlButton>
    )]
  );
};

/**
 * Controls Component
 * @param props { newOnly: boolean }
 */
const Controls: FC<IProps> = ({ newOnly }) => {
  const onNewGame = () => '';
  const onRestart = () => '';
  const onSeeOriginal = () => '';

  return (
    <ControlsWrapper newOnly={newOnly}>
      <ControlButton
        action={onNewGame}
        type={ControlType.NEW_GAME}
      >
        New Game
      </ControlButton>
      {!newOnly ? secondaryControls(onRestart, onSeeOriginal) : null}
    </ControlsWrapper>
  );
};

Controls.defaultProps = {
  newOnly: false,
};

export default Controls;
