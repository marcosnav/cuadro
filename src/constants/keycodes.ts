import { SwipeDirection } from './directions';

/**
 * KeyCodes to match for keypress events
 */
export enum KeyCodes {
  DOWN = 40,
  LEFT = 37,
  RIGHT = 39,
  UP = 38,
  ENTER = 13,
}

/**
 * Map key code with swipe direction
 */
export const KeySwipeMap: { [key: number]: SwipeDirection } = {
  40: SwipeDirection.DOWN,
  37: SwipeDirection.LEFT,
  39: SwipeDirection.RIGHT,
  38: SwipeDirection.UP,
};
