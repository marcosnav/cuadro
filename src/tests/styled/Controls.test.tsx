import { mount } from 'enzyme';
import 'jest-styled-components';
import React from 'react';
import { ThemeConsumer } from 'styled-components';
import Controls from './../../components/styled/Controls';
import { theme } from './../../constants';

describe('<Controls/>', () => {
  beforeAll(() => {
    ThemeConsumer._currentValue = theme;
  });

  test('renders properly', () => {
    const component = mount(<Controls />);
    expect(component).toHaveStyleRule('background', theme.WHITE);
    expect(component).toHaveStyleRule('padding', '0 16px');
    expect(component).toHaveStyleRule('width', '300px');
    expect(component).toMatchSnapshot();
  });

  describe('with prop: new-only', () => {
    test('shows only the new game option', () => {
      const component = mount(<Controls newOnly={true} />);
      expect(component).toHaveStyleRule('padding', '0 10px');
      expect(component).toHaveStyleRule('width', '77px');
      expect(component).toMatchSnapshot();
    });
  });
});
