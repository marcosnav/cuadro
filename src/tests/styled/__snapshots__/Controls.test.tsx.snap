// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Controls/> renders properly 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  border-radius: 3px;
  box-shadow: 0 4px 8px 3px rgba(0,0,0,0.2);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 68px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 26px;
  padding: 0 16px;
  -webkit-transition: all 0.6s ease;
  transition: all 0.6s ease;
  width: 300px;
}

<Controls>
  <section
    className="c0"
  />
</Controls>
`;

exports[`<Controls/> with prop: new-only shows only the new game option 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  border-radius: 3px;
  box-shadow: 0 4px 8px 3px rgba(0,0,0,0.2);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 68px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: calc(100vh - 360px);
  padding: 0 10px;
  -webkit-transition: all 0.6s ease;
  transition: all 0.6s ease;
  width: 77px;
}

<Controls
  newOnly={true}
>
  <section
    className="c0"
  />
</Controls>
`;
